reveal-core/src/main/kotlin/com/svenjacobs/reveal/RevealOverlay.kt

```diff
@@ -8,7 +8,7 @@ import androidx.compose.ui.unit.IntRect
 import androidx.compose.ui.unit.LayoutDirection
 
 enum class RevealOverlayAlignment {
-	Start, Top, End, Bottom,
+	Start, Top, End, Bottom, TopNoCenter, BottomNotCenter
```

Comment:
We should probably allow all possible combinations here:

```kotlin
enum class RevealOverlayAlignment {
	StartTop,
	StartCenter,
	StartBottom,
	TopStart,
	TopCenter,
	TopEnd,
	EndTop,
	EndCenter,
	EndBottom,
	BottomStart,
	BottomCenter,
	BottomEnd,
}
```

Reply:
Looking at how Compose handles alignments internally, I'm not quite satisfied how Reveal solves this currently. I think a major refactoring is required. Ideally we should use the alignment types already provided by Compose.

Reply:
Maybe we could use the existing [2D alignments](https://developer.android.com/reference/kotlin/androidx/compose/ui/Alignment) of Compose to align the overlay content inside the reveal area and provide an additional `anchor` attribute  (`Start, Top, End, Bottom`) to move (offset) the content to one side of the area?

Reply:
Would that still provide the option for the overlays to be aligned respective the the parent? For my purposes, I need the overlay to be aligned to the top of bottom nav bar item reveal areas for all the items in the bar but the horizontal position cannot change. Specifically, I need to implement the exact overlay design above and have it aligned identically for all the items in the nav bar

Reply:
So what you want to achieve is placing the overlay above the revealable but expanding to the full width of the screen?
If we have something like `no center` we should also provide this for `Start` and `End`. I'm trying to think of a generic solution that is useful for any use case.

Reply:
I'm looking at adding some additional alignment and anchor options, do you think something like this could work? (I've only implemented vertical alignment currently)
```kotlin
enum class VerticalRevealOverlayAlignment {
	Top, Bottom
}

enum class VerticalRevealOverlayAnchor {
	ParentStart, ParentEnd, ParentCenter, RevealableStart, RevealableEnd, RevealableCenter
}

override fun Modifier.align(
	verticalAlignment: VerticalRevealOverlayAlignment,
	anchor: VerticalRevealOverlayAnchor
): Modifier = this.then(
	Modifier.layout { measurable, constraints ->
		val placeable = measurable.measure(constraints)
		layout(constraints.maxWidth, constraints.maxHeight) {
			val actualAnchor = when {
				layoutDirection == LayoutDirection.Rtl
					&& anchor == VerticalRevealOverlayAnchor.ParentStart -> VerticalRevealOverlayAnchor.ParentEnd
				layoutDirection == LayoutDirection.Rtl
					&& anchor == VerticalRevealOverlayAnchor.ParentEnd -> VerticalRevealOverlayAnchor.ParentStart
				layoutDirection == LayoutDirection.Rtl
					&& anchor == VerticalRevealOverlayAnchor.RevealableStart -> VerticalRevealOverlayAnchor.RevealableEnd
				layoutDirection == LayoutDirection.Rtl
					&& anchor == VerticalRevealOverlayAnchor.RevealableEnd -> VerticalRevealOverlayAnchor.RevealableStart
				else -> anchor
			}

			when {
				verticalAlignment == VerticalRevealOverlayAlignment.Top
					&& actualAnchor == VerticalRevealOverlayAnchor.ParentStart -> {
					placeable.place(
						x = 0,
						y = revealRect.top - placeable.height
					)
				}
				verticalAlignment == VerticalRevealOverlayAlignment.Top
					&& actualAnchor == VerticalRevealOverlayAnchor.ParentCenter -> {
					placeable.place(
						x = (constraints.maxWidth / 2) - (placeable.width / 2),
						y = revealRect.top - placeable.height
					)
				}
				verticalAlignment == VerticalRevealOverlayAlignment.Top
					&& actualAnchor == VerticalRevealOverlayAnchor.ParentEnd -> {
					placeable.place(
						x = constraints.maxWidth - (placeable.width / 2),
						y = revealRect.top - placeable.height
					)
				}
				verticalAlignment == VerticalRevealOverlayAlignment.Top
					&& actualAnchor == VerticalRevealOverlayAnchor.RevealableStart -> {
					placeable.place(
						x = revealRect.left,
						y = revealRect.top - placeable.height
					)
				}
				verticalAlignment == VerticalRevealOverlayAlignment.Top
					&& actualAnchor == VerticalRevealOverlayAnchor.RevealableCenter -> {
					placeable.place(
						x = revealRect.left + (revealRect.width - placeable.width) / 2,
						y = revealRect.top - placeable.height
					)
				}
				verticalAlignment == VerticalRevealOverlayAlignment.Top
					&& actualAnchor == VerticalRevealOverlayAnchor.RevealableEnd -> {
					placeable.place(
						x = revealRect.right,
						y = revealRect.top - placeable.height
					)
				}

				verticalAlignment == VerticalRevealOverlayAlignment.Bottom
					&& actualAnchor == VerticalRevealOverlayAnchor.ParentStart -> {
					placeable.place(
						x = 0,
						y = revealRect.bottom
					)
				}
				verticalAlignment == VerticalRevealOverlayAlignment.Bottom
					&& actualAnchor == VerticalRevealOverlayAnchor.ParentCenter -> {
					placeable.place(
						x = (constraints.maxWidth / 2) - (placeable.width / 2),
						y = revealRect.bottom
					)
				}
				verticalAlignment == VerticalRevealOverlayAlignment.Bottom
					&& actualAnchor == VerticalRevealOverlayAnchor.ParentEnd -> {
					placeable.place(
						x = constraints.maxWidth - (placeable.width / 2),
						y = revealRect.bottom
					)
				}
				verticalAlignment == VerticalRevealOverlayAlignment.Bottom
					&& actualAnchor == VerticalRevealOverlayAnchor.RevealableStart -> {
					placeable.place(
						x = revealRect.left,
						y = revealRect.bottom
					)
				}
				verticalAlignment == VerticalRevealOverlayAlignment.Bottom
					&& actualAnchor == VerticalRevealOverlayAnchor.RevealableCenter -> {
					placeable.place(
						x = revealRect.left + (revealRect.width - placeable.width) / 2,
						y = revealRect.bottom
					)
				}
				verticalAlignment == VerticalRevealOverlayAlignment.Bottom
					&& actualAnchor == VerticalRevealOverlayAnchor.RevealableEnd -> {
					placeable.place(
						x = revealRect.right,
						y = revealRect.bottom
					)
				}
			}
		}
	}
)
```

Reply:
I just pushed a new commit that adds 2 new methods to API for horizontal and vertical alignment of the overlay with anchor options relative to both the parent and the revealable. Let me know if that's similar to what you were thinking

Reply:
Thank you. I probably have time next week to review this.

Reply:
Thanks! I will let me team know I'm putting my work implementing Reveal on hold pending your review

Reply:
Digging a bit deeper into this topic I noticed that the alignment / placement isn't ideal, either. For instance if an overlay item has a lot of text and is placed to the left or right side of a revealable, the item isn't properly wrapped and goes beyond the visible screen. I guess there is more work to do. Sorry for the delay.

If you're really blocked by this what you could do in the meantime is fork the project, apply your changes and work with this version until I have fixed the layout issues.

Reply:
I agree, having overlays placed off the screen is not ideal. I think this could be pretty easily resolved by doing a min/max to constrain the placement within the window. Would still leave some wonky behaviour if an overlay was positioned above a revealable that was already at the top of the screen, for example, but I think there's a reasonable level of responsibility on the developer to pick reasonable placements. I'll look at adding some constrain bounding.

Reply:
Added some constraints the placement, take a look and let me know what you think!

Reply:
> Added some constraints the placement, take a look and let me know what you think!

I'm not sure if adding `min` or `max` to the `place` function is really the solution here. `place` just places the element on the X and Y axis and doesn't say anything about the size, unless I'm mistaken. I guess the solution lies somewhere in the values passed to the `layout` function above. But I have to further investigate custom layouts in Compose to be really sure.

Furthermore I would like to get rid of the large, imperative `when` statement and put the placement logic in instances of the alignment classes, just like Compose's `Alignment` is implemented.

Reply:
Makes sense, I agree there's definitely room for improvement in the alignment code. Would you consider merging some of my changes in to address additional layout options in the mean time? My project is somewhat time sensitive and these changes would allow us to roll out the feature, even if the underlying implementation is refactored in a future iteration.

Reply:
> Makes sense, I agree there's definitely room for improvement in the alignment code. Would you consider merging some of my changes in to address additional layout options in the mean time? My project is somewhat time sensitive and these changes would allow us to roll out the feature, even if the underlying implementation is refactored in a future iteration.

Would it be okay for you to fork the project and apply your needed changes there, just as a temporary solution? I would like to avoid releasing a version with changes to the API that I know will be completely overhauled anyway soon.

Reply:
I mean, you already forked it 😅 so what you could do is build that version and put the library in the [libs](https://sebhastian.com/android-studio-libs-folder/) folder of your Android project.

Reply:
Yep I think we'll go that route in the meantime. Thanks for your input!